src/coreclr/vm/jitinterface.cpp

```diff
@@ -5066,7 +5101,7 @@ void CEEInfo::getCallInfo(
     }
     else
     {
-        if (!exactType.IsTypeDesc() && !pTargetMD->IsArray())
+        if (!exactType.IsTypeDesc() && !pTargetMD->IsArray() && !pTargetMD->IsDynamicMethod())
```

Comment:
@jkotas CoreCLR is asserting here on `DynamicMethod` without adding a check here, is this the right solution? With it we can emit direct calls to them but they're still not inlined cause VM says they're `DONTINLINE`.

Reply:
What's the assert?

Dynamic methods can call each other today. This code is able to handle that. How is the case that you are introducing different?

Reply:
> What's the assert?

It's the null return assert inside of this if.

> How is the case that you are introducing different?

The `MethodTable*` from the DynamicMethod is different from the one this method obtains from the JIT which in turn gets it from `getMethodClass(replacementMethod)`. It seems like `getMethodClass` causes the pointer to be normalized to the real one while the VM here has a special DynamicMethod one which causes `GetExactDeclaringType` to fail.

Reply:
I am not following. `DynamicMethodTable` and `MethodTable` are unrelated types. `DynamicMethodTable` should never show up in the JIT/EE interface calls.

Reply:
> I am not following. `DynamicMethodTable` and `MethodTable` are unrelated types. `DynamicMethodTable` should never show up in the JIT/EE interface calls.

I mean, the JIT is passing the proper class handle, but the method table stored in the dynamic MethodDesc is a custom one with the debug name for it being `dynamicClass`, which makes `GetExactDeclaringType` fail cause that seems to have no relation to the parent type.

Reply:
Doing:
```cpp
if (pTargetMD->IsDynamicMethod())
{
    exactType = pTargetMD->GetMethodTable();
}
else if (!exactType.IsTypeDesc() && !pTargetMD->IsArray())
```

instead also seems to work, not sure what is more correct here. The current version leaves exactType as the class provided to the DynamicMethod while this makes it `dynamicClass`.

Reply:
Ok, so the problem is that `getMethodClass` returns the owning class handle for dynamic methods, but `CORINFO_RESOLVED_TOKEN` is expected to have the actual class handle for dynamic methods. 

I think it would be best to fix it in the place that fills in `CORINFO_RESOLVED_TOKEN` to have the actual class handle. It will make the dynamic methods path introduced by this PR work the same way as the existing paths taken by the dynamic methods.

Reply:
> Ok, so the problem is that `getMethodClass` returns the owning class handle for dynamic methods, but `CORINFO_RESOLVED_TOKEN` is expected to have the actual class handle for dynamic methods.
> 
> I think it would be best to fix it in the place that fills in `CORINFO_RESOLVED_TOKEN` to have the actual class handle. It will make the dynamic methods path introduced by this PR work the same way as the existing paths taken by the dynamic methods.

How would I grab that from the JIT? [This](https://github.com/dotnet/runtime/pull/109679/files#diff-266282d463447c6984d3223b87f05a0cc94197441c10c204a8f89af0d3d8a372R213) is where I grab that currently with `getMethodClass`.

Reply:
Also to be honest, this seems to me a bit like CoreCLR is leaking implementation details of itself to the JIT, for me it'd make more sense if it always returned the handle of the owner class to the JIT and then mapped back to it's dynamic one on it's end reliably. Otherwise we're polluting the JIT with behaviours specific to one VM.

Reply:
I agree that CoreCLR is leaking the implementation details here, but I have different opinion about what the correct behavior should be.

The dynamic method type owner is optional and it is only meant to be used for visibility checks. All visibility checks should be done on the EE side. I am not sure why the JIT ever needs to see the type owner. I think it would make more sense for the JIT to always see the special MethodTable that the dynamic methods are attached to.

> Would you suggest to add it to the getMethodFromDelegate JIT-EE api?

I would think so. It may also help you with communicating precise type to the JIT. I think `MethodDesc*` may not always have the precise type in case of shared generic code - notice that `Delegate.GetMethodImpl` does extra work to figure out the precise type for generic types https://github.com/dotnet/runtime/blob/main/src/coreclr/System.Private.CoreLib/src/System/Delegate.CoreCLR.cs#L169-L170.

Reply:
> I would think so. It may also help you with communicating precise type to the JIT. I think `MethodDesc*` may not always have the precise type in case of shared generic code - notice that `Delegate.GetMethodImpl` does extra work to figure out the precise type for generic types https://github.com/dotnet/runtime/blob/main/src/coreclr/System.Private.CoreLib/src/System/Delegate.CoreCLR.cs#L169-L170.

I can't find a case where this is an issue, the JIT seems to always have the proper generic context to resolve stuff so I wonder if this is just a reflection only issue?

If not, could you assist me with handling this cause I couldn't figure out how to replicate this logic in VMs?